numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +336,93 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!intern_strings())
+        return NULL;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs))
+        return NULL;
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (PyDict_CheckExact(kwargs) &&
+            PyDict_Contains(kwargs, npy_arrayfunction_str_like))
+        PyDict_DelItem(kwargs, npy_arrayfunction_str_like);
+
+    return array_implement_array_function_internal(
+        implementation, public_api, relevant_args, args, kwargs);
+}
+
+
+/*
+ * Implements the __array_function__ protocol for a C function. Added as an extension
+ * to NEP-18 in an effort to bring NEP-35 to life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    PyObject *relevant_args;
+    PyObject *numpy_module = NULL, *public_api = NULL;
+    PyObject *result = NULL;
+
+    if (kwargs == NULL)
+        return NULL;
+
+    if (!intern_strings())
+        goto cleanup;
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (PyDict_CheckExact(kwargs) &&
```

Comment:
Probably, I don't know if there's a case where kwargs could be (probably by accident) something else, so I decided to take the better-safe-than-sorry approach. Are you suggesting we simply ignore it and rely only on https://github.com/numpy/numpy/pull/16935/files/580067f634c056ee251d05adec363a2b19b1b2b9#diff-785bfb041ce7cc5030614c05f873a44aR381-R382 ?

Reply:
Hmmm, I think we can just assume it is a dict or NULL (in `array_implement_c_array_function`). In the python exposed version, `kwargs` is passed in from python. That is from autogenerated code, but at least its a bit more plausible for it being a non-dict. Which would be a big error (which python probably checks later).  In principle, it may actually be best to do a `PyDict_ExactCheck` in the Python version and set a manual error.

Reply:
I'm assuming it's a dict or NULL then, done in latest commit.